tools/prow-config-updater/updater.go

```diff
@@ -0,0 +1,189 @@
+		dryrun:            dryrun,
+	}
+	// If the PR is created by the Prow bot, we should be confident to blindly update production Prow configs.
+	if *pr.User.Name == config.ProwBotName {
+		if err := cu.updateProw(config.ProdProwEnv); err != nil {
+			return fmt.Errorf("error updating production Prow configs: %v", err)
+		}
+	} else {
+		// Check if we need staging process for updating the config files.
+		if cu.needsStaging() {
+			if err := cu.startStaging(); err != nil {
+				// Best effort, won't fail the process if the comment fails.
+				cu.githubcommenter.commentOnStagingFailure(*pr.Number, err)
+				return fmt.Errorf("error running Prow staging process: %v", err)
+			} else {
+				// Best effort, won't fail the process if the comment fails.
+				cu.githubcommenter.commentOnStagingSuccess(*pr.Number, err)
+			}
+			// TODO(chizhg): create an auto-merge pull request to update production Prow configs.
+		} else {
+			if err := cu.updateProw(config.ProdProwEnv); err != nil {
+				return fmt.Errorf("error updating production Prow configs: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// Update Prow with the changed config files and send message for the update result.
+func (cu *ConfigUpdater) updateProw(env config.ProwEnv) error {
+	updatedFiles, err := cu.doProwUpdate(env)
+	prnumber := cu.prnumber
+	if err == nil {
+		// Best effort, won't fail the process if the comment fails.
+		cu.githubcommenter.commentOnUpdateConfigsSuccess(prnumber, config.ProdProwEnv, updatedFiles)
+	} else {
+		// Best effort, won't fail the process if the comment fails.
+		cu.githubcommenter.commentOnUpdateConfigsFailure(prnumber, config.StagingProwEnv, updatedFiles, err)
+	}
+	return err
+}
+
+// Decide if we need staging process by checking the changed files.
+func (cu *ConfigUpdater) needsStaging() bool {
+	// If any key config files for staging Prow are changed, the staging process will be needed.
+	fs := collectRelevantFiles(cu.files, config.StagingProwKeyConfigPaths)
+	return len(fs) != 0
+}
+
+// Start running the staging process to update and test staging Prow.
+func (cu *ConfigUpdater) startStaging() error {
+	// Update staging Prow.
+	if err := cu.updateProw(config.StagingProwEnv); err != nil {
+		return fmt.Errorf("error updating staging Prow configs: %v", err)
+	}
+	// Create pull request in the fork repository for the testing of staging Prow.
+	fpr, err := cu.githubmainhandler.createForkPullRequest()
+	if err != nil {
+		return fmt.Errorf("error creating pull request in the fork repository: %v", err)
+	}
+	// Create comment on the fork pull request to get it tested by staging Prow and merged.
+	forkprnumber := *fpr.Number
+	if err = cu.githubcommenter.commentToMergePullRequest(forkprnumber); err != nil {
+		return fmt.Errorf("error creating comment on the fork pull request: %v", err)
+	}
+	// Wait for the fork pull request to be automatically merged by staging Prow.
+	if err := cu.githubmainhandler.waitForForkPullRequestMerged(*fpr.Number); err != nil {
```

Comment:
Also, want to do something in Go fashion: kick off a goroutine waiting for this event

Reply:
I'm using a library in `k8s.io/apimachinery/pkg/util/wait` to poll the status periodically, and it's using goroutine and channel under the hood. Or do you mean something else?

Reply:
I see, that's what I need